### PR TITLE
fix: Duplicate accounts 

### LINKED ISSUE #9823: [Bug]: Duplicate Account Issue in MetaMask 7.23.0 on iPhone 14
### Describe the bug

After updating MetaMask to version 7.23.0 on my iPhone 14, the same account appears duplicated in the app. When attempting to open the duplicated account, it displays the original account with the same address. A friend with an iPhone 14 Pro has experienced the exact same issue.

### Expected behavior

I expected to see only one instance of each account and be able to access each account individually without duplication.

### Screenshots/Recordings

Attached is a screenshot showing the duplicated accounts. The original account is highlighted with a red
314C098B-0977-4515-926D-BBE277213B74
 frame.

### Steps to reproduce

1. Update MetaMask to version 7.23.0 on an iPhone 14 or iPhone 14 Pro.
2. Open MetaMask and check the list of accounts.
3. Notice the duplicated accounts.
4. Try to open the duplicated account, which displays the original account instead.

### Error messages or log output

```shell
No specific error messages are shown.
```

### Version

7.23.0

### Build type

None

### Device

iPhone 14, iPhone 14 Pro

### Operating system

iOS

### Additional context

This issue affects multiple users with similar devices and the same version of MetaMask.

### Severity

_No response_

### PR DESCRIPTION
**Pull request by owencraston**
## **Description**
This PR addresses the issue of duplicate accounts showing up in the wallet. This bug was introduced when [the accounts controller was integrated](https://github.com/MetaMask/metamask-mobile/pull/8759) but was only visible to users in [this PR](https://github.com/MetaMask/metamask-mobile/pull/9733) when we started rendering account information in the accounts list. The is that when we ran migration `036.ts`, we simply added the all of the identities from the preferences controller into the AccountsController. However the accounts controller also consumes addresses from the KeyringController and those addresses are in lowercase. The result would be two accounts with the same address, one in checksum format and one in lowercase format. This would result in duplicates when rendering the UI. 

This PR addresses this issue in two ways...
1. for users who are using <= v7.19.0
    - I have patched migration 36 to ensure that the addresses that are added are lowercase. 
2. For users who have are using version >= 7.20.0 and are already effected by this bug
    - I have created a new migration to deduplicate the accounts in the accounts controller and ensure that we are only storing lowercase addresses in state.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/9823

## QA Builds
- [Current branch](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c68d80ea-dbaa-4e81-9018-fe9757cab397)
- [7.18.0](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/011ac0b3-b1d8-4e53-9595-ea0f739d4d10)
- [7.24.0](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a5c54c50-8382-45ad-b203-687b66273c4e)

### Path 1: User has duplicate accounts already
1. Install [7.18.0](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/011ac0b3-b1d8-4e53-9595-ea0f739d4d10) and import a wallet with multiple addresses already
2. Install [7.24.0](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a5c54c50-8382-45ad-b203-687b66273c4e) and upgrade app.
3. EXPECTED: You should see duplicate accounts when you open the accounts list. If you had 5 accounts previously there will be should now be 10.
4. Install [the current branch QA build](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c68d80ea-dbaa-4e81-9018-fe9757cab397) and upgrade.
5. EXPECT: The duplicate accounts should be removed (back to 5 accounts from 10)
6. EXPECT: The previously selected address should still be selected
7. EXPECT: The account names should be in tact and in chronological order if they are the default account name.

### Path 2: User has not seen duplicate accounts yet
this occurs for users who have not upgraded from 7.19.0 yet
1. Install [7.18.0](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/011ac0b3-b1d8-4e53-9595-ea0f739d4d10) and import a wallet with multiple addresses already
2. Install [the current branch QA build](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c68d80ea-dbaa-4e81-9018-fe9757cab397) and upgrade.
3. EXPECT: All of the account should appear in the accounts list without duplicates
4. EXPECT: The previously selected address should be the same
5. EXPECT: The account names should be the same

### **Before**

7.18.0

After Upgrading to 7.24.0

### **After**

Upgrading to this branch

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/components/hooks/useAccounts/useAccounts.ts [hunk 1]
```diff
@@ -52,6 +53,20 @@ const useAccounts = ({
   const internalAccounts = useSelector(selectInternalAccounts);
   const selectedInternalAccount = useSelector(selectSelectedInternalAccount);
 
+  console.log(
```
Comment (owencraston) [inline app/components/hooks/useAccounts/useAccounts.ts:56]: to be removed after testing

**Comment by owencraston** [comment]
QA builds: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/3dab86ec-30d3-498a-b9d7-c82577044e71

File: app/store/migrations/044.ts
Comment (Cal-L) [inline app/store/migrations/044.ts:1]: Need to import this and add to list of migrations in migrations/index.ts

File: app/store/migrations/044.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  const addressMap: Record<string, string> = {};
+
+  for (const [, account] of Object.entries(internalAccounts)) {
+    const lowerCaseAddress = account.address.toLowerCase();
+    if (addressMap[lowerCaseAddress]) {
+      const existingAccount = mergedAccounts[addressMap[lowerCaseAddress]];
+      existingAccount.metadata = {
+        ...existingAccount.metadata,
+        ...account.metadata,
+        name: existingAccount.metadata.name || account.metadata.name,
```
Comment (Cal-L) [inline app/store/migrations/044.ts:62]: This seems like it would allow dupes to overwrite the account names.

**Review by Cal-L — COMMENTED** [review]
Let's provide vids for the two cases

- Upgrading from 7.19.0 -> this branch showing that dupes are no longer created
- Upgrading from 7.19.0 -> 7.23 to get into a state where dupes appear. Then upgrade to this branch showing that dupes are removed.

File: app/store/migrations/044.ts
Comment (owencraston) [reply app/store/migrations/044.ts:1]: 🤦

File: app/store/migrations/044.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      addressMap[lowerCaseAddress] = newId;
+      mergedAccounts[newId] = {
+        ...account,
+        address: lowerCaseAddress,
+        id: newId,
+      };
+    }
+  }
+
+  const newSelectedAccount = addressMap[selectedAddress] || selectedAccount;
```
Comment (Cal-L) [inline app/store/migrations/044.ts:78]: This should not fallback to selectedAccount since more likely than not, the ID no longer exists

Comment (owencraston) [reply app/store/migrations/044.ts]: fixed, in the latest commit. The default selected account is now the first account in the list.

File: app/store/migrations/044.ts (hunk 2 above)
Comment (owencraston) [reply app/store/migrations/044.ts]: I modified the code to prefer custom account names using the `isDefaultAccountName` util function. The logic is as follows...

```
function deriveAccountName(existingName: string, currentName: string): string {
  const isExistingNameDefault = isDefaultAccountName(existingName);
  const isCurrentNameDefault = isDefaultAccountName(currentName);

  return isExistingNameDefault && !isCurrentNameDefault
    ? currentName
    : existingName;
}
```

File: app/store/migrations/044.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  const selectedAddress =
+    internalAccounts[selectedAccount]?.address.toLowerCase();
+
+  const mergedAccounts: Record<string, InternalAccount> = {};
+  const addressMap: Record<string, string> = {};
+
+  for (const [, account] of Object.entries(internalAccounts)) {
+    const lowerCaseAddress = account.address.toLowerCase();
+    if (addressMap[lowerCaseAddress]) {
```
Comment (Cal-L) [inline app/store/migrations/044.ts:67]: nit: abstract and reuse `addressMap[lowerCaseAddress]` as `accountID`

File: app/store/migrations/044.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  return isExistingNameDefault && !isCurrentNameDefault
+    ? currentName
+    : existingName;
+}
+
+function mergeInternalAccounts(state: Record<string, any>) {
+  const accountsController: AccountsControllerState =
+    state.engine.backgroundState.AccountsController;
+  const internalAccounts = accountsController.internalAccounts.accounts;
+  const selectedAccount = accountsController.internalAccounts.selectedAccount;
```
Comment (Cal-L) [inline app/store/migrations/044.ts:57]:
```suggestion
  const selectedAccountId = accountsController.internalAccounts.selectedAccount;
```

File: app/store/migrations/044.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+  const selectedAddress =
+    internalAccounts[selectedAccount]?.address.toLowerCase();
```
Comment (Cal-L) [inline app/store/migrations/044.ts:60]:
```suggestion
    internalAccounts[selectedAccountId]?.address.toLowerCase();
```

File: app/store/migrations/044.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      addressMap[lowerCaseAddress] = newId;
+      mergedAccounts[newId] = {
+        ...account,
+        address: lowerCaseAddress,
+        id: newId,
+      };
+    }
+  }
+
+  const newSelectedAccount =
```
Comment (Cal-L) [inline app/store/migrations/044.ts:91]:
```suggestion
  const newSelectedAccountId =
```

File: app/store/migrations/044.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    addressMap[selectedAddress] ||
+    Object.keys(mergedAccounts)[0] || // Default to the first account in the list
+    selectedAccount;
+
+  accountsController.internalAccounts.accounts = mergedAccounts;
+  accountsController.internalAccounts.selectedAccount = newSelectedAccount;
```
Comment (Cal-L) [inline app/store/migrations/044.ts:97]:
```suggestion
  accountsController.internalAccounts.selectedAccount = newSelectedAccountId;
```

File: app/store/migrations/044.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        address: lowerCaseAddress,
+        id: newId,
+      };
+    }
+  }
+
+  const newSelectedAccount =
+    addressMap[selectedAddress] ||
+    Object.keys(mergedAccounts)[0] || // Default to the first account in the list
+    selectedAccount;
```
Comment (Cal-L) [inline app/store/migrations/044.ts:94]: Remove this fallback since it's dead code and will never be hit

File: app/core/Engine.ts [hunk 10]
```diff
@@ -574,6 +574,9 @@ class Engine {
       messenger: accountsControllerMessenger,
       state: initialState.AccountsController ?? defaultAccountsControllerState,
     });
+
+    accountsController.updateAccounts();
```
Comment (owencraston) [inline app/core/Engine.ts:578]: This needs to get called once after the user has been logged in. This was missed during the initial integration and is leading to outdated account types. [See this comment](https://github.com/MetaMask/metamask-mobile/blob/86076ac96d85df80d5f3a8bd73c6c30cc8837aae/app/store/migrations/036.ts#L120-L123). Is there a better place to put this @Cal-L @tommasini ?

**Review by Cal-L — APPROVED** [review]
LGTM

File: app/core/Engine.ts (hunk 10 above)
Comment (owencraston) [reply app/core/Engine.ts]: For the sake of speed and reducing complexity we are not going to introduce this in this patch.

**Comment by owencraston** [comment]
Tested with qa builds a well as locally and found that it was working. No duplicate state and all of the accounts controller state is intact.

**Pull request merged**